Lib/_pylong.py

```diff
@@ -210,6 +210,91 @@ def inner(a, b):
     w5pow = compute_powers(len(s), 5, DIGLIM)
     return inner(0, len(s))
 
+# Asymptotically faster version, using the C decimal module. See
+# comments at the end of the file. This uses decimal arithmetic to
+# convert from base 10 to base 256. The latter is just a string of
+# bytes, which CPython can convert very efficiently to a Python int.
+
+import math
+_LOG_10_BASE_256 = math.log(10, 256)
+del  math
+
+def _dec_str_to_int_inner(s):
+    BYTELIM = 512
+    D = decimal.Decimal
+    result = bytearray()
+
+    def inner(n, w):
+        if w <= BYTELIM:
+            # XXX Stefan Pochmann discovered that, for 1024-bit ints,
+            # `int(Decimal)` took 2.5x longer than `int(str(Decimal))`.
+            # So simplify this code to the former if/when that gets
+            # repaired.
+            result.extend(int(str(n)).to_bytes(w)) # big-endian default
+            return
+        w2 = w >> 1
+        if 0:
+            # This is maximally clear, but "too slow". `decimal`
+            # division is asymptotically fast, but we have no way to
+            # tell it to reuse the high-precision reciprocal it computes
+            # for pow256[w2], so it has to recompute it over & over &
+            # over again :-(
+            hi, lo = divmod(n, pow256[w2][0])
+        else:
+            p256, recip = pow256[w2]
+            # The integer part will have about half the digits of n. So
+            # only need that much precision, plus some guard digits.
+            ctx.prec = (n.adjusted() >> 1) + 8
+            hi = +n * +recip # unary `+` chops back to ctx.prec digits
+            ctx.prec = decimal.MAX_PREC
+            hi = hi.to_integral_value() # lose the fractional digits
+            lo = n - hi * p256
+            # Because we've been uniformly rounding down, `hi` is a
+            # lower bound on the correct quotient.
+            assert lo >= 0
+            # Adjust quotient up if needed. It usually isn't. In random
+            # testing, the loop body entered about one in 100 thousand
+            # cases. I never saw it need more than one iteration.
+            count = 0
+            while lo >= p256:
+                count += 1
+                # If the assert fails, chances are decent we're sooooo
+                # far off it may seem to run forever otherwise - the
+                # error analysis was fatally flawed in this case.
+                assert count < 10, (count, w, str(n))
+                lo -= p256
+                hi += 1
+            # The assert should always succeed, but way too slow to keep
+            # enabled.
+            #assert hi, lo == divmod(n, pow256[w2][0])
+        inner(hi, w - w2)
+        inner(lo, w2)
+
+    w = int(len(s) * _LOG_10_BASE_256) + 1
+    with decimal.localcontext(_unbounded_dec_context) as ctx:
```

Comment:
Does it create a copy of `_unbounded_dec_context` or returns a reference to `_unbounded_dec_context`?

Reply:
It creates a copy, which the `with:` exit handler throws away. `_unbounded_dec_context` is not altered.

Reply:
Nice. So this code is not only thead-safe, but re-entrant.